### ISSUE TITLE
ValueError: Attempting to unscale FP16 gradients. on V100 with --fp16

### ISSUE THREAD

**Issue by YerongLi** [haotian-liu/LLaVA#294]
### When did you clone our code?

I cloned the code base after 5/1/23

### Describe the issue

Issue:
I found `--bits 4` or `--bits 8` works on A100 
while quantization with `--bits 8` and `--fp16 True` does not work on V100 GPU. (with torch.autocast("cuda") enabled)
```
if __name__ == "__main__":
    with torch.autocast("cuda"):
        train()
```
Command:
```
#!/bin/bash

WEIGHT_VERSION=v1-1
PROMPT_VERSION=v1
MODEL_VERSION="7b"

python llava/train/train.py \
    --bits 8 \
    --lora_enable True \
    --model_name_or_path checkpoints/vicuna-7b-v1.1 \
    --version $PROMPT_VERSION \
    --data_path ./playground/data/llava_instruct_150k.json \
    --image_folder /scratch/yerong/Multimodal-GPT/data/coco/train2017 \
    --vision_tower openai/clip-vit-large-patch14 \
    --pretrain_mm_mlp_adapter ./checkpoints/mm_projector/LLaVA-7b-pretrain-projector-v0-CC3M-595K-original_caption.bin \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end True \
    --fp16 True \
    --output_dir ./checkpoints/deepspeed_llava-$MODEL_VERSION-$WEIGHT_VERSION-finetune_lora \
    --num_train_epochs 3 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 2e-4 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --dataloader_num_workers 4 \
    --report_to wandb
```
**LOG**
```
equenceClassification model from a BertForSequenceClassification model).
Formatting inputs...Skip in lazy mode
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
wandb: Currently logged in as: yerong. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.15.5
wandb: Run data is saved locally in /scratch/yerong/LLaVA/wandb/run-20230721_081035-slaltz5i
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run easy-darkness-41
wandb: _ View project at https://wandb.ai/yerong/huggingface
wandb:  View run at https://wandb.ai/yerong/huggingface/runs/slaltz5i
  0%|                                                                                                                              | 0/473136 [00:00<?, ?it/s]/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py:322: UserWarning: MatMul8bitLt: inputs will be cast from torch.float32 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Traceback (most recent call last):
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 935, in <module>
    train()
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 914, in train
    trainer.train()
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1645, in train
    return inner_training_loop(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1987, in _inner_training_loop
    self.accelerator.clip_grad_norm_(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/accelerate/accelerator.py", line 1893, in clip_grad_norm_
    self.unscale_gradients()
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/accelerate/accelerator.py", line 1856, in unscale_gradients
    self.scaler.unscale_(opt)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 284, in unscale_
    optimizer_state["found_inf_per_device"] = self._unscale_grads_(optimizer, inv_scale, found_inf, False)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 212, in _unscale_grads_
    raise ValueError("Attempting to unscale FP16 gradients.")
ValueError: Attempting to unscale FP16 gradients.
```

Screenshots:
![screenshot](https://github.com/haotian-liu/LLaVA/assets/13112023/6c61e6d2-928b-46d9-8fa7-d8bd0f6d6af8)

**Comment by haotian-liu** [comment]
Unfortunately, I do not currently have access to V100s. And from the screenshot, I am assuming that this is an issue with training with QLora.

I will double check the QLora this coming week.

If you have any updates w.r.t. this, please also kindly share. Thanks!

**Comment by haotian-liu** [comment]
@YerongLi 

I have updated the QLora support. Please pull the latest repo and reinstall the packages using `pip install -e .`

See example script [here](https://github.com/haotian-liu/LLaVA/blob/main/scripts/finetune_qlora.sh).

**Comment by YerongLi** [comment]
Could you test the `--fp16` option on A100? you script uses bf16

with latest code, I am getting new error: `RuntimeError: expected scalar type Half but found Float`
```
wandb: _ View project at https://wandb.ai/yerong/huggingface                                                                                                  
wandb:  View run at https://wandb.ai/yerong/huggingface/runs/423v20eu
  0%|                                                                                                                              | 0/473136 [00:00<?, ?it/s]
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes
 in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that exce
ssive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(                                                                                                                          
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py:322: UserWarning: MatMul8bitLt: inputs will be cast from to
rch.float32 to float16 during quantization                                                                                                                    
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Traceback (most recent call last):                                                                                                                            
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 931, in <module>
    train()
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 909, in train
    trainer.train()
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1645, in train
    return inner_training_loop(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1938, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 2770, in training_step
    self.accelerator.backward(loss)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/accelerate/accelerator.py", line 1819, in backward
    self.scaler.scale(loss).backward(**kwargs)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/autograd/__init__.py", line 200, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/autograd/function.py", line 274, in apply
    return user_fn(self, *args)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 157, in backward
    torch.autograd.backward(outputs_with_grad, args_with_grad)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/autograd/__init__.py", line 200, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/autograd/function.py", line 274, in apply
    return user_fn(self, *args)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py", line 480, in backward
    grad_A = torch.matmul(grad_output, CB).view(ctx.grad_shape).to(ctx.dtype_A) 
RuntimeError: expected scalar type Half but found Float
```
![screenshot](https://github.com/haotian-liu/LLaVA/assets/13112023/52963460-ae54-4127-b930-b2499d8906a9)

**Comment by YerongLi** [comment]
These two has to be false
```
    --bf16 False \
    --tf32 False \
```

**Comment by YerongLi** [comment]
> Qlora works with bf16 computing, but for GPU without bf16 computing it does not work.

**Comment by haotian-liu** [comment]
@YerongLi We need to enable fp16 and disable bf16 both in DeepSpeed config. I have tried to change it auto in https://github.com/haotian-liu/LLaVA/commit/6b3079ea5da535599c3e56344c283adc319e532f, so that moving forward we just need to make sure fp16 is enabled and bf16 is disabled in command line, on GPUs like V100s that do not support bf16.

I have tried the fp16 qlora and it works on 3090, by adding `--fp16 True --bf16 False --tf32 False`. Can you help verify this on V100? Also, can you help also verify that `--bf16 True` will still raise an error on V100?

Thanks.

**Comment by YerongLi** [comment]
Are there any ways to run `train.py` directly without deepspeed?

**Comment by YerongLi** [comment]
Good to know deepspeed works. Are there any ways to run `train.py` directly without deepspeed?
> @YerongLi We need to enable fp16 and disable bf16 both in DeepSpeed config. I have tried to change it auto in [6b3079e](https://github.com/haotian-liu/LLaVA/commit/6b3079ea5da535599c3e56344c283adc319e532f), so that moving forward we just need to make sure fp16 is enabled and bf16 is disabled in command line, on GPUs like V100s that do not support bf16.
> 
> I have tried the fp16 qlora and it works on 3090, by adding `--fp16 True --bf16 False --tf32 False`. Can you help verify this on V100? Also, can you help also verify that `--bf16 True` will still raise an error on V100?
> 
> Thanks.

**Comment by YerongLi** [comment]
Yeah, because single GPU works with qlora

**Comment by haotian-liu** [comment]
Simply remove deepspeed config from training command, and run command starting with `python` would work.

Also, just want to point out that QLoRA works with deepspeed as well.

**Comment by YerongLi** [comment]
Okay, let me try again.

**Comment by YerongLi** [comment]
```
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes
 in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that exce
ssive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(                                                                                                                          
wandb: Currently logged in as: yerong. Use `wandb login --relogin` to force relogin                                                
wandb: wandb version 0.15.8 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade                                          
wandb: Tracking run with wandb version 0.15.5                                                                                                                 
wandb: Run data is saved locally in /scratch/yerong/LLaVA/wandb/run-20230804_144504-jbe7nnj3
wandb: Run `wandb offline` to turn off syncing.                                                                                                               
wandb: Syncing run stoic-frost-51                                              
wandb: _ View project at https://wandb.ai/yerong/huggingface
wandb:  View run at https://wandb.ai/yerong/huggingface/runs/jbe7nnj3
  0%|                                                                                                                              | 0/157712 [00:00<?, ?it/s]
/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes
 in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that exce
ssive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
Traceback (most recent call last):
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 931, in <module>
    train()
  File "/scratch/yerong/LLaVA/llava/train/train.py", line 909, in train
    trainer.train()
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1539, in train
    return inner_training_loop(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/transformers/trainer.py", line 1867, in _inner_training_loop
    self.accelerator.clip_grad_norm_(
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/accelerate/accelerator.py", line 1925, in clip_grad_norm_
    self.unscale_gradients()
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/accelerate/accelerator.py", line 1888, in unscale_gradients
    self.scaler.unscale_(opt)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 284, in unscale_
    optimizer_state["found_inf_per_device"] = self._unscale_grads_(optimizer, inv_scale, found_inf, False)
  File "/scratch/yerong/.conda/envs/llava/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 212, in _unscale_grads_
    raise ValueError("Attempting to unscale FP16 gradients.")
ValueError: Attempting to unscale FP16 gradients.
```

> Simply remove deepspeed config from training command, and run command starting with `python` would work.
> 
> Also, just want to point out that QLoRA works with deepspeed as well.

I pip install -e all over and tested the following script:
```
#!/bin/bash

WEIGHT_VERSION=v1-1
PROMPT_VERSION=v1
MODEL_VERSION="7b"

python llava/train/train.py \
    --lora_enable True \
    --bits 4 \
    --model_name_or_path checkpoints/vicuna-7b-v1.1 \
    --version $PROMPT_VERSION \
    --data_path ./playground/data/llava_instruct_150k.json \
    --image_folder /scratch/yerong/Multimodal-GPT/data/coco/train2017 \
    --vision_tower openai/clip-vit-large-patch14 \
    --pretrain_mm_mlp_adapter ./checkpoints/mm_projector/LLaVA-7b-pretrain-projector-v0-CC3M-595K-original_caption.bin \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --fp16 True \
    --bf16 False \
    --output_dir ./checkpoints/llava-$MODEL_VERSION-finetune_lora \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --dataloader_num_workers 4 \
    --report_to wandb
```

**Comment by haotian-liu** [comment]
I also found the same thing, but do not find an obvious way to resolve this. Do you need to actually train the model with a single GPU? Or you just want to debug?

**Comment by YerongLi** [comment]
Train it on single GPU will work on A100 with qlora (And it is affordable), while on V100 it does not work because of this `--fp16` bug

**Comment by haotian-liu** [comment]
Not sure what the exact cause is, but if you want to train with a single GPU, a simple workaround is to train on a single GPU using DeepSpeed :)

**Comment by YerongLi** [comment]
Yeah not sure what is happening, I mean deepspeed works 4x slower than the train_mem.py (bf16) on A100, not sure whether it will work with fp16 V100.

**Comment by haotian-liu** [comment]
Which configuration are you comparing with? +/- DeepSpeed, +bf16,+train_mem, +qlora?

**Comment by YerongLi** [comment]
deepspeed works indeed :)

> Not sure what the exact cause is, but if you want to train with a single GPU, a simple workaround is to train on a single GPU using DeepSpeed :)
